### PR TITLE
#31: Created token classification UDF

### LINKED ISSUE #31: Create Token Classification UDF
# Background
- Token classification labels each tokens in a given text.
- The common token classification tasks are  Named Entity Recognition (NER)  and  Part-of-Speech (PoS) tagging.
- The main purpose of the NER is to classify tokens in a text, such as dates, individuals and places
- The main purpose of the PoS is to label tokens in a text such as verbs, nouns, and punctuation marks.
- Transformers API provides [AutoModelForTokenClassification](https://huggingface.co/docs/transformers/task_summary#named-entity-recognition) class to perform token classification

# Acceptance Criteria
- Create UDF for token classification
- Add unit/integration tests

### PR DESCRIPTION
**Pull request by umitbuyuksahin**
Fixes #31

### CODE REVIEW

File: exasol_transformers_extension/udfs/models/named_entity_recognition_udf.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        :param model_df: The dataframe to be predicted
+
+        :return: List of dataframe includes prediction details
+        """
+
+        text_data = list(model_df['text_data'])
+        results = self.last_created_pipeline(text_data)
+        results = results if type(results[0]) == list else [results]
+
+        columns = ["word", "entity", "score"]
```
Comment (tkilias) [inline exasol_transformers_extension/udfs/models/named_entity_recognition_udf.py:142]: do we have a word position? because the word itself is not uniquely identified

Comment (tkilias) [reply exasol_transformers_extension/udfs/models/named_entity_recognition_udf.py:142]: Output rows of UDFs can be resorted in the database, for that reason, we need to provide a position if want to represent a specific order of rows

File: exasol_transformers_extension/udfs/models/named_entity_recognition_udf.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
+import torch
+import pandas as pd
+import transformers
+from typing import Tuple, List, Union, Dict
+from exasol_transformers_extension.deployment import constants
+from exasol_transformers_extension.utils import device_management, \
+    dataframe_operations, bucketfs_operations
+
+
+class NamedEntityRecognitionUDF:
```
Comment (tkilias) [inline exasol_transformers_extension/udfs/models/named_entity_recognition_udf.py:10]: Is this only NER or also TokenClassification

**Review by tkilias — CHANGES_REQUESTED** [review]

File: exasol_transformers_extension/udfs/models/named_entity_recognition_udf.py (hunk 1 above)
Comment (umitbuyuksahin) [reply exasol_transformers_extension/udfs/models/named_entity_recognition_udf.py:142]: Yes wenm can provide position information

File: exasol_transformers_extension/udfs/models/named_entity_recognition_udf.py (hunk 2 above)
Comment (umitbuyuksahin) [reply exasol_transformers_extension/udfs/models/named_entity_recognition_udf.py:10]: This is only NER

File: exasol_transformers_extension/udfs/models/token_classification_udf.py [hunk 3]
```diff
@@ -137,10 +142,16 @@ def _predict_model(self, model_df: pd.DataFrame) -> List[pd.DataFrame]:
...
         text_data = list(model_df['text_data'])
-        results = self.last_created_pipeline(text_data)
+        aggregation_strategy = model_df['aggregation_strategy'].iloc[0]
+        results = self.last_created_pipeline(
+            text_data, aggregation_strategy=aggregation_strategy)
         results = results if type(results[0]) == list else [results]
 
-        columns = ["start", "end", "word", "entity_group", "score"]
+        if aggregation_strategy == "none":
+            columns = ["start", "end", "word", "entity", "score"]
```
Comment (tkilias) [inline exasol_transformers_extension/udfs/models/token_classification_udf.py:151]: I think, if you use here entity, that won work with the rename below

Comment (tkilias) [reply exasol_transformers_extension/udfs/models/token_classification_udf.py:151]:
```
            result_df = result_df[columns].rename(
                columns={
                    "start": "start_pos",
                    "end": "end_pos",
                    "entity_group": "entity"})
```

File: exasol_transformers_extension/udfs/models/token_classification_udf.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+                (batch_df['bucketfs_conn'] == bucketfs_conn) &
+                (batch_df['sub_dir'] == sub_dir)]
+
+            current_model_key = (bucketfs_conn, sub_dir, model_name)
+            if self.last_loaded_model_key != current_model_key:
+                self.set_cache_dir(model_df)
+                self.clear_device_memory()
+                self.load_models(model_name)
+                self.last_loaded_model_key = current_model_key
+                self.last_used_top_k = None
```
Comment (tkilias) [inline exasol_transformers_extension/udfs/models/token_classification_udf.py:70]: I think, we don't need top_k for token classification

File: exasol_transformers_extension/udfs/models/token_classification_udf.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+        :param model_df: The dataframe to be predicted
+
+        :return: List of dataframe includes prediction details
+        """
+
+        text_data = list(model_df['text_data'])
+        aggregation_strategy = model_df['aggregation_strategy'].iloc[0]
+        results = self.last_created_pipeline(
+            text_data, aggregation_strategy=aggregation_strategy)
```
Comment (tkilias) [inline exasol_transformers_extension/udfs/models/token_classification_udf.py:147]: can you really give the aggregation_strategy into the call, the documentation looks like, as that you would need to specify the strategy at pipeline creation

**Review by tkilias — CHANGES_REQUESTED** [review]

File: exasol_transformers_extension/udfs/models/token_classification_udf.py (hunk 4 above)
Comment (umitbuyuksahin) [reply exasol_transformers_extension/udfs/models/token_classification_udf.py:70]: This variable is leftover. We dont have topk for this task

File: exasol_transformers_extension/udfs/models/token_classification_udf.py (hunk 5 above)
Comment (umitbuyuksahin) [reply exasol_transformers_extension/udfs/models/token_classification_udf.py:147]: Yes, api document is not good enough I think.
Check tests in the without_db where different strategies are called.

**Comment by tkilias** [comment]
I saw this warnings, be careful with them, this might indicate that your dataframe operation, don't have the effect you expect, this specifically import if you modify the dataframe

> 
>   /home/runner/work/transformers-extension/transformers-extension/exasol_transformers_extension/udfs/models/text_generation_udf.py:170: SettingWithCopyWarning: 
>   A value is trying to be set on a copy of a slice from a DataFrame.
>   Try using .loc[row_indexer,col_indexer] = value instead
>   
>   See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
>     model_df['generated_text'] = generated_texts
> 
> tests/unit_tests/udfs/test_token_classification.py: 14 warnings
>   /home/runner/work/transformers-extension/transformers-extension/exasol_transformers_extension/udfs/models/token_classification_udf.py:72: SettingWithCopyWarning: 
>   A value is trying to be set on a copy of a slice from a DataFrame
>   
>   See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
>     model_df.fillna(self.default_aggregation_strategy, inplace=True)

File: tests/integration_tests/without_db/udfs/test_token_classification_udf.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    ("on CPU with single input with NULL aggregation", None, 1, None),
+    ("on CPU with single input with max aggregation", None, 1, "max"),
+    ("on GPU with batch input with none aggregation", 0, 3, "none"),
+    ("on GPU with batch input with NULL aggregation", 0, 3, None),
+    ("on GPU with batch input with max aggregation", 0, 3, "max"),
+    ("on GPU with single input with none aggregation", 0, 1, "none"),
+    ("on GPU with single input with NULL aggregation", 0, 1, None),
+    ("on GPU with single input with max aggregation", 0, 1, "max")
+])
+def test_token_classification_udf(
```
Comment (tkilias) [inline tests/integration_tests/without_db/udfs/test_token_classification_udf.py:58]: I think, another test with multiple aggregation_strategies would be helpful here, to check the assumption that we can specify the aggregation_strategy during call

Comment (umitbuyuksahin) [reply tests/integration_tests/without_db/udfs/test_token_classification_udf.py:57]: We are already able to introduce this parameter in the call; we would get an error if we couldn't.

Comment (tkilias) [reply tests/integration_tests/without_db/udfs/test_token_classification_udf.py:57]: This test is for testing, if we can change the strategy, sometimes, you can set it ones, but not change it later.

**Review by tkilias — APPROVED** [review]
